File: problem39/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        if (root == null)
            return new ArrayList<>();
        List<Integer> leftBranch = new ArrayList<>();
        leftBranch = rightSideView(root.left);

        List<Integer> rightBranch = new ArrayList<>();
        rightBranch = rightSideView(root.right);

        List<Integer> finalBranch = new ArrayList<>();
        finalBranch.add(root.val);
        
        if (rightBranch.size() < leftBranch.size()) {
            finalBranch.addAll(rightBranch);
			// finalBranch.addAll(finalBranch.size(), leftBranch.subList(rightBranch.size(), leftBranch.size()));
			// or
            for (int i = rightBranch.size(); i < leftBranch.size(); i++) {
                finalBranch.add(leftBranch.get(i));
            }
        } else
            finalBranch.addAll(rightBranch);
        return finalBranch;
    }
}

File: problem39/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        # Return empty list if root is null
        if not root: return []

        # DFS left & right branches and store theirs lists
        left = self.rightSideView(root.left)
        right = self.rightSideView(root.right)

        # Remember sizes of left & right lists
        l = 0 if not left else len(left)
        r = 0 if not right else len(right)

        # If right list is longer, return root and right list
        if r >= l:
            return [root.val] + right

        ''' Otherwise return root + right list +
            the part of left list that is longer '''
        return [root.val] + right + left[r:]
